File: payload.go
package main

import (
	. "github.com/linux-surface/iptsd/protocol"
)

func IptsPayloadHandleInput(ipts *IptsContext) error {
	payload, err := ipts.Protocol.ReadPayload()
	if err != nil {
		return err
	}

	for i := uint32(0); i < payload.Frames; i++ {
		frame, err := ipts.Protocol.ReadPayloadFrame()
		if err != nil {
			return err
		}

		switch frame.Type {
		case IPTS_PAYLOAD_FRAME_TYPE_STYLUS:
			err = IptsStylusHandleInput(ipts, frame)
			break
		case IPTS_PAYLOAD_FRAME_TYPE_TOUCH:
			err = IptsTouchHandleInput(ipts, frame)
			break
		default:
			// ignored
			err = ipts.Protocol.Skip(frame.Size)
			break
		}

		if err != nil {
			return err
		}
	}

	return nil
}

File: payload.c
// SPDX-License-Identifier: GPL-2.0-or-later

#include "context.h"
#include "protocol.h"
#include "stylus.h"
#include "touch.h"
#include "syscall.h"

int iptsd_payload_handle_input(struct iptsd_context *iptsd,
		struct ipts_data *header)
{
	int pos = 0;
	struct ipts_payload *payload = (struct ipts_payload *)header->data;

	for (int i = 0; i < payload->frames; i++) {
		int ret = 0;
		struct ipts_payload_frame *frame =
			(struct ipts_payload_frame *)&payload->data[pos];

		switch (frame->type) {
		case IPTS_PAYLOAD_FRAME_TYPE_STYLUS:
			ret = iptsd_stylus_handle_input(iptsd, frame);
			break;
		case IPTS_PAYLOAD_FRAME_TYPE_TOUCH:
			ret = iptsd_touch_handle_input(iptsd, frame);
			break;
		}

		if (ret < 0) {
			iptsd_err(ret, "Failed to handle payload frame");
			return ret;
		}

		pos = pos + sizeof(struct ipts_payload_frame) + frame->size;
	}

	return 0;
}

